akka-stream/src/main/scala/akka/stream/scaladsl/FlowWithContext.scala

```diff
@@ -0,0 +1,119 @@
+    })
+
+  def filter(pred: Out ⇒ Boolean): Repr[Ctx, Out] =
+    collect { case e if pred(e) ⇒ e }
+
+  def filterNot(pred: Out ⇒ Boolean): Repr[Ctx, Out] =
+    collect { case e if !pred(e) ⇒ e }
+
+  def grouped(n: Int): Repr[immutable.Seq[Ctx], immutable.Seq[Out]] =
+    via(flow.grouped(n).map { elsWithContext ⇒
+      val (els, ctxs) = elsWithContext.unzip
+      (els, ctxs)
+    })
+
+  def sliding(n: Int, step: Int = 1): Repr[immutable.Seq[Ctx], immutable.Seq[Out]] =
+    via(flow.sliding(n, step).map { elsWithContext ⇒
+      val (els, ctxs) = elsWithContext.unzip
+      (els, ctxs)
+    })
+
+  def mapConcat[Out2](f: Out ⇒ immutable.Iterable[Out2]): Repr[Ctx, Out2] = statefulMapConcat(() ⇒ f)
+
+  def statefulMapConcat[Out2](f: () ⇒ Out ⇒ immutable.Iterable[Out2]): Repr[Ctx, Out2] = {
+    val fCtx: () ⇒ ((Out, Ctx)) ⇒ immutable.Iterable[(Out2, Ctx)] = { () ⇒ elWithContext ⇒
+      val (el, ctx) = elWithContext
+      f()(el).map(o ⇒ (o, ctx))
+    }
+    via(flow.statefulMapConcat(fCtx))
+  }
+
+  def mapContext[Ctx2](f: Ctx ⇒ Ctx2): Repr[Ctx2, Out] =
+    via(flow.map { case (e, ctx) ⇒ (e, f(ctx)) })
+
+  def endContextPropagation: Prov[Ctx, Out]
+
+  private[akka] def flow[T, C]: Flow[(T, C), (T, C), NotUsed] = Flow[(T, C)]
+}
+
+/**
+ * API MAY CHANGE
+ */
+@ApiMayChange
+object FlowWithContext {
+  def apply[Ctx, In]: FlowWithContext[Ctx, In, Ctx, In, akka.NotUsed] = {
+    val under = Flow[(In, Ctx)]
+    new FlowWithContext[Ctx, In, Ctx, In, akka.NotUsed](under, under.traversalBuilder, under.shape)
+  }
+  def from[CI, I, CO, O, M](flow: Flow[(I, CI), (O, CO), M]) = new FlowWithContext(flow, flow.traversalBuilder, flow.shape)
+}
+
+/**
+ * API MAY CHANGE
+ */
+@ApiMayChange
+final class FlowWithContext[-CtxIn, -In, +CtxOut, +Out, +Mat](
+  underlying:                    Flow[(In, CtxIn), (Out, CtxOut), Mat],
+  override val traversalBuilder: LinearTraversalBuilder,
+  override val shape:            FlowShape[(In, CtxIn), (Out, CtxOut)]
+) extends FlowWithContextOps[CtxOut, Out, Mat] with Graph[FlowShape[(In, CtxIn), (Out, CtxOut)], Mat] {
+
+  override def withAttributes(attr: Attributes): Repr[CtxOut, Out] = new FlowWithContext(underlying, traversalBuilder.setAttributes(attr), shape)
+
+  override type Repr[+C, +O] = FlowWithContext[CtxIn @uncheckedVariance, In @uncheckedVariance, C, O, Mat @uncheckedVariance]
+  override type Prov[+C, +O] = Flow[(In @uncheckedVariance, CtxIn @uncheckedVariance), (O, C), Mat @uncheckedVariance]
+
+  override def via[Ctx2, Out2, Mat2](viaFlow: Graph[FlowShape[(Out, CtxOut), (Out2, Ctx2)], Mat2]): Repr[Ctx2, Out2] = from(underlying.via(viaFlow))
+
+  def to[Mat2](sink: Graph[SinkShape[(Out, CtxOut)], Mat2]): Sink[(In, CtxIn), Mat] = underlying.toMat(sink)(Keep.left)
+
+  def toMat[Mat2, Mat3](sink: Graph[SinkShape[(Out, CtxOut)], Mat2])(combine: (Mat, Mat2) ⇒ Mat3): Sink[(In, CtxIn), Mat3] = underlying.toMat(sink)(combine)
+
+  override def endContextPropagation: Prov[CtxOut, Out] = underlying
```

Comment:
Do we even need to support that here? We could say that you can only use `FlowWithContext` on a `SourceWithContext`.

It would be nice if we could require that `startContextPropagation` and `stopContextPropagation` must always be paired.

Otherwise, we should rename it, perhaps to `asFlow`.

Reply:
We definitely need a `endContextPropagation` on both `SourceWithContext` and `FlowWithContext`. I'm not sure how you want to require a pairing of start and stop, but I think `asFlow` is probably a better, simpler name.

Reply:
> I'm not sure how you want to require a pairing of start and stop

Pairing doesn't work for the `FlowWithContext`, that's why I would change the name here to `asFlow`. For `SourceWithContext` pairing should work because you never start with a `SourceWithContext` but always with a `Source`. So it's

```scala
source
  .startContextPropagation
  .map(...)
  ...
  .endContextPropagation
  .runWith(...)
```

Reply:
ok make sense, looks good.